File: threatlens/threatlens-streamlit/threatlens/modules/taxonomy.py
"""Threat Taxonomy Engine (TT-FR-01..06).

Structured, hierarchical classification of phishing threats.
"""
from __future__ import annotations

from threatlens.config import TAXONOMY


def get_taxonomy() -> dict[str, str]:
    """TT-FR-01/05: return the taxonomy as {category: description}."""
    return dict(TAXONOMY)


def classify(sample: dict) -> str:
    """TT-FR-03: assign a sample to a taxonomy category.

    TODO: implement rule/ML-based classification. Placeholder returns the
    sample's existing label or the most common category.
    """
    return sample.get("label", "Credential Theft")


def get_distribution() -> dict[str, float]:
    """Share of the dataset per category (for the Campaign View bars)."""
    # TODO: compute from real labeled data.
    return {
        "Credential Theft": 41,
        "Fake Login Page": 27,
        "Business Email Compromise": 16,
        "Financial Scam": 10,
        "Malware Delivery": 6,
    }

File: threatlens/threatlens-web/js/modules/taxonomy.js
/* Threat Taxonomy Engine (TT-FR-01/03/05).
   Structured classification + keyword-based category assignment. */
TL.taxonomy = (function () {
  return {
    getTaxonomy() { return Object.assign({}, TL.config.TAXONOMY); },      // TT-FR-05

    /** classify(sample) -> category label (TT-FR-03). Assumes phishing input. */
    classify(sample) {
      const t = ((sample.subject || "") + " " + (sample.body || "")).toLowerCase();
      const url = (sample.url || "").toLowerCase();
      if (/wire transfer|are you available|process a payment|discreet/.test(t)) return "Business Email Compromise";
      if (/\.zip|attachment|download the|\.exe|document has been shared/.test(t + url)) return "Malware Delivery";
      if (/password|verify|credential|confirm your account/.test(t)) return "Credential Theft";
      if (/invoice|payment|refund|prize|reward|won|claim/.test(t)) return "Financial Scam";
      if (/sign ?in|login|log ?in|unusual sign/.test(t + url)) return "Fake Login Page";
      return "Credential Theft";
    },

    /** get_distribution() -> {category: percent} across current phishing samples. */
    getDistribution(samples) {
      const counts = {};
      samples.filter(s => s.label && s.label !== "Legitimate").forEach(s => {
        counts[s.label] = (counts[s.label] || 0) + 1;
      });
      const total = Object.values(counts).reduce((a, b) => a + b, 0) || 1;
      const out = {};
      Object.keys(counts).sort((a, b) => counts[b] - counts[a]).forEach(k => {
        out[k] = Math.round((counts[k] / total) * 100);
      });
      return out;
    }
  };
})();
